### PR TITLE
Non numeric categories colormap

### PR DESCRIPTION
**Pull request by squirrelo**
This adds the ability to do colormaps for mixed string and numeric columns. Useful if you have a numeric column with a few string types for missing data, or for highlighting missing values in what should otherwise be a numeric column.

An error alert is given if there are 1 or 0 numeric values, otherwise it does this:

(this gif is outdated):
coloring

### CODE REVIEW

File: examples/build.py [hunk 1]
```diff
@@ -43,10 +45,10 @@ def listify(a):
 coords = np.random.randn(N, 10).tolist()
 pct_var = (1/np.exp(np.arange(10))).tolist()
 
-md_headers = ['SampleID', 'DOB']
+md_headers = ['SampleID', 'DOB', 'Strings']
 metadata = []
 for _id in coords_ids:
-    metadata.append([_id, ''.join(sample(categories, 1))])
+    metadata.append([_id, ''.join(sample(categories, 1)), ''.join(choice(ascii_letters) for x in range(10))])
```
Comment (ElDeveloper) [inline examples/build.py]: Mind keeping this at 80 chars?

File: emperor/support_files/js/color-view-controller.js [hunk 2]
```diff
@@ -132,16 +132,26 @@ define([
           var data = decompViewDict.setCategory(attributes, scope.setPlottableAttributes, category);
 
           if (scaled) {
-            scope.setSlickGridDataset({});
-            scope.$gridDiv.hide();
+            //build plottables as all non-numerics
```
Comment (ElDeveloper) [inline emperor/support_files/js/color-view-controller.js]: Mind adding a leading space in this and all the other comments?

File: emperor/support_files/js/color-view-controller.js [hunk 3]
```diff
@@ -132,16 +132,26 @@ define([
...
+            var nonNumeric = uniqueVals.filter(function(v) { return isNaN(v)});
```
Comment (ElDeveloper) [inline emperor/support_files/js/color-view-controller.js]: I think you should be able to filter passing just the isNaN function.

File: emperor/support_files/js/color-view-controller.js (hunk 2 above)
Comment (ElDeveloper) [inline emperor/support_files/js/color-view-controller.js]: I think this whole block about determining non-numeric values and their respective plottables belongs in a separate function that we should be able to test, the edge cases here sound like a nightmare to have untested.

File: emperor/support_files/js/color-view-controller.js [hunk 4]
```diff
@@ -132,16 +132,26 @@ define([
...
+            var plotList = data.filter(function(x) {
+              return $.inArray(x.category, nonNumeric) !== -1;
+            });
+            var plottables = [];
+            for (var i = 0; i < plotList.length; i++) {
+              plottables = plottables.concat(plotList[i].plottables);
+            }
+            // Set SlickGrid for color of non-numeric values and show color bar for rest
+            scope.setSlickGridDataset([{category: 'Non-numeric values', value:'#64655d', plottables: plottables}]);
+            scope.$gridDiv.css('height', '30px');
```
Comment (ElDeveloper) [inline emperor/support_files/js/color-view-controller.js]: In general, we don't really want to have a hardcoded height, this makes it hard to maintain a consistent size after resizing callbacks. Why can't we use the hide/show methods?

File: emperor/support_files/js/color-view-controller.js [hunk 5]
```diff
@@ -193,71 +204,154 @@ define([
    * continuous).
    */
   ColorViewController.getColorList = function(values, map, discrete, scaled) {
-    var colors = {}, numColors = values.length - 1, counter = 0;
-    var interpolator, min, max;
+    var scope = this;
```
Comment (ElDeveloper) [inline emperor/support_files/js/color-view-controller.js]: This is a class method, how is it that you need `scope` here? Is this vestigial from the refactoring? Also just noticed that `scaled` is being redefined below i.e. ti is part of the function's arguments and then is re-declared with var, since you are making changes to this file, would you mind changing that?

File: emperor/support_files/js/color-view-controller.js [hunk 6]
```diff
@@ -193,71 +204,154 @@ define([
...
+    if (values.length == 1 && discrete === false) {
       colors[values[0]] = chroma.brewer[map][0];
       return [colors, gradientSVG];
     }
 
-    if (discrete === false) {
+    //Call helper function to create the required colormap type
+    if (discrete) {
+      colors = ColorViewController.getDiscreteColors(values, map);
+    } else if (scaled) {
```
Comment (ElDeveloper) [inline emperor/support_files/js/color-view-controller.js]: Would you mind putting all the else blocks in a new line?

File: emperor/support_files/js/color-view-controller.js [hunk 7]
```diff
@@ -193,71 +204,154 @@ define([
...
+      var info = ColorViewController.getScaledColors(values, map);
+      colors = info[0];
+      gradientSVG = info[1];
+    } else {
+      colors = ColorViewController.getInterpolatedColors(values, map)
+    }
+    return [colors, gradientSVG];
```
Comment (ElDeveloper) [inline emperor/support_files/js/color-view-controller.js]: Can you update the docstring?

File: emperor/support_files/js/color-view-controller.js [hunk 8]
```diff
@@ -193,71 +204,154 @@ define([
...
+   *
+   * This function will generate a list of coloring values depending on the
+   * coloring scheme that the system is currently using (discrete or
+   * continuous).
+   *
+   * Defaults to use qiime discrete colors if there's no map passed in.
+   *
+   */
+  ColorViewController.getDiscreteColors = function(values, map) {
+    var map = map || 'discrete-coloring-qiime';
```
Comment (ElDeveloper) [inline emperor/support_files/js/color-view-controller.js]: You are redefining map here.

File: emperor/support_files/js/color-view-controller.js [hunk 9]
```diff
@@ -193,71 +204,154 @@ define([
...
+
+  /**
+   *
+   * Retrieve a discrete color.
+   *
+   * @param {values} list of objects to generate a color for, usually a
+   * category in a given metadata column.
+   * @param {map} string, name of the discrete color map to use.
+   *
+   * This function will generate a list of coloring values depending on the
```
Comment (ElDeveloper) [inline emperor/support_files/js/color-view-controller.js]: Just noticed this block is out of date, the map is an argument and does not depend on the "system".

File: emperor/support_files/js/color-view-controller.js [hunk 10]
```diff
@@ -193,71 +204,154 @@ define([
...
-    if (discrete === false) {
+    //Call helper function to create the required colormap type
+    if (discrete) {
+      colors = ColorViewController.getDiscreteColors(values, map);
+    } else if (scaled) {
+      var info = ColorViewController.getScaledColors(values, map);
+      colors = info[0];
+      gradientSVG = info[1];
+    } else {
+      colors = ColorViewController.getInterpolatedColors(values, map)
```
Comment (ElDeveloper) [inline emperor/support_files/js/color-view-controller.js]: Can you add a semicolon at the end of this line?

File: emperor/support_files/js/color-view-controller.js [hunk 11]
```diff
@@ -193,71 +204,154 @@ define([
...
+   *
+   * This function will generate a list of coloring values depending on the
+   * coloring scheme that the system is currently using (discrete or
+   * continuous).
+   *
+   * Defaults to use viridis colormap if there's no map passed in.
+   *
+   */
+  ColorViewController.getScaledColors = function(values, map, nanColor) {
+    var map = map || 'viridis';
```
Comment (ElDeveloper) [inline emperor/support_files/js/color-view-controller.js]: Can you remove the `var` in this and the next line?

File: emperor/support_files/js/color-view-controller.js [hunk 12]
```diff
@@ -193,71 +204,154 @@ define([
...
+  /**
+   *
+   * Retrieve a scaled color set with scale bar.
+   *
+   * @param {values} list of objects to generate a color for, usually a
+   * category in a given metadata column.
+   * @param {map} string, name of the discrete color map to use.
+   *
+   * This function will generate a list of coloring values depending on the
+   * coloring scheme that the system is currently using (discrete or
```
Comment (ElDeveloper) [inline emperor/support_files/js/color-view-controller.js]: This block is also out of date, can you remove it? Also, can you add a `@return` section?

File: emperor/support_files/js/color-view-controller.js [hunk 13]
```diff
@@ -193,71 +204,154 @@ define([
...
+   */
+  ColorViewController.getScaledColors = function(values, map, nanColor) {
+    var map = map || 'viridis';
+    var nanColor = nanColor || '#64655d';
+    map = chroma.brewer[map];
+
+    // Get list of only numeric values, error if none
+    var numericValues = values.filter(function(v) { return !isNaN(v)});
+    if (numericValues.length < 2) {
+      alert('less than 2 numeric values found, can not create scale for category!');
```
Comment (ElDeveloper) [inline emperor/support_files/js/color-view-controller.js]: Can we remove the `alert` calls from these functions? These shouldn't care about the logic of the UI, these should be merely helpers.

Comment (ElDeveloper) [inline emperor/support_files/js/color-view-controller.js]: Just to be more specific, what would be helpful is if we had appropriate try/catch blocks on the consumers of these functions.

File: emperor/support_files/js/color-view-controller.js [hunk 14]
```diff
@@ -193,71 +204,154 @@ define([
...
+    //build the SVG showing the gradient of colors for values
+    var mid = (min + max) / 2;
+    var step = (max - min) / 100;
+    var stopColors = [];
+    for (var s = min; s <= max; s += step) {
+      stopColors.push(interpolator(s).hex());
     }
-
+    var gradientSVG = '<defs><linearGradient id="Gradient" x1="0" x2="1" y1="1" y2="1">';
+    for (pos in stopColors) {
```
Comment (ElDeveloper) [inline emperor/support_files/js/color-view-controller.js]: Just noticed this, can you change this iteration to use the indices instead of the `in` keyword?

File: emperor/support_files/js/color-view-controller.js [hunk 15]
```diff
@@ -193,71 +204,154 @@ define([
...
+   * Defaults to use viridis colormap if there's no map passed in.
+   *
+   */
+  ColorViewController.getScaledColors = function(values, map, nanColor) {
+    var map = map || 'viridis';
+    var nanColor = nanColor || '#64655d';
+    map = chroma.brewer[map];
+
+    // Get list of only numeric values, error if none
+    var numericValues = values.filter(function(v) { return !isNaN(v)});
```
Comment (ElDeveloper) [inline emperor/support_files/js/color-view-controller.js]: I'm a little bit worried about this whole block, we are doing multiple passes over the same array of values doing slightly different things at each time, is there a way we could do a single pass with a helper function?

File: emperor/support_files/js/color-view-controller.js [hunk 16]
```diff
@@ -193,71 +204,154 @@ define([
...
+    var map = map || 'viridis';
+    map = chroma.brewer[map];
+
+    var total = values.length;
+    var counter = 0;
+    interpolator = chroma.bezier([map[0], map[3], map[4], map[5], map[8]]);
+    var colors = {};
+    for(var i = 0; i < values.length; i++) {
+      colors[values[i]] = interpolator(counter / total).hex();
+      counter++;
```
Comment (ElDeveloper) [inline emperor/support_files/js/color-view-controller.js]: How is `counter` different from `i`?

File: emperor/support_files/js/color-view-controller.js [hunk 17]
```diff
@@ -193,71 +204,154 @@ define([
...
+   *
+   * This function will generate a list of coloring values depending on the
+   * coloring scheme that the system is currently using (discrete or
+   * continuous).
+   *
+   * Defaults to use viridis colormap if there's no map passed in.
+   *
+   */
+  ColorViewController.getInterpolatedColors = function(values, map) {
+    var map = map || 'viridis';
```
Comment (ElDeveloper) [inline emperor/support_files/js/color-view-controller.js]: Can you remove the `var`?

File: emperor/support_files/js/color-view-controller.js [hunk 18]
```diff
@@ -193,71 +204,154 @@ define([
...
 
   /**
+   *
+   * Retrieve an interpolatd color set.
+   *
+   * @param {values} list of objects to generate a color for, usually a
+   * category in a given metadata column.
+   * @param {map} string, name of the discrete color map to use.
+   *
+   * This function will generate a list of coloring values depending on the
```
Comment (ElDeveloper) [inline emperor/support_files/js/color-view-controller.js]: Can you remove this block and add the `@return` section?

File: tests/javascript_tests/test_color_view_controller.js [hunk 19]
```diff
@@ -126,7 +126,7 @@ requirejs([
 
     test("Test ColorViewController.getColorList exceptions", function(){
       var five;
-      five = [0, 1, 2, 3, 4];
+      five = ['0', 'string1', 'string2', 'string3', 'string4'];
```
Comment (ElDeveloper) [inline tests/javascript_tests/test_color_view_controller.js]: I'm skeptical that we are testing all the separate functions by just adding a new test case 😆 can you add the appropriate test cases?

**Comment by ElDeveloper** [comment]
Overall functionality looks good, though there are a few problems I noticed:

1.- Grid gets hidden after switching between tabs: if you turn your coloring scheme into a continuous "coloring" and then switch to another tab, then come back, you are going to see that the grid is now clipped:

screen shot 2016-05-07 at 5 23 43 pm

The problem persists even after changing the "continuous" coloring:

screen shot 2016-05-07 at 5 26 08 pm

Also, in testing this feature, I noticed that the labels for the colormap overlap when the precision is high:

screen shot 2016-05-07 at 5 28 12 pm

File: emperor/support_files/js/color-view-controller.js (hunk 3 above)
Comment (squirrelo) [inline emperor/support_files/js/color-view-controller.js]: I can, but I wanted to keep this consistent between all the filter functions.

File: emperor/support_files/js/color-view-controller.js (hunk 4 above)
Comment (squirrelo) [inline emperor/support_files/js/color-view-controller.js]: I need to have a specific height for a single row of the slickgrid, so you can still change the color of the non-numeric values. I will dig into slickgrid to see if I can pull that height dynamically, unless you know the command.

File: emperor/support_files/js/color-view-controller.js (hunk 5 above)
Comment (squirrelo) [inline emperor/support_files/js/color-view-controller.js]: all vestigial of refactoring and moving from protected to class method.

File: emperor/support_files/js/color-view-controller.js (hunk 7 above)
Comment (squirrelo) [inline emperor/support_files/js/color-view-controller.js]: done

File: emperor/support_files/js/color-view-controller.js (hunk 9 above)
Comment (squirrelo) [inline emperor/support_files/js/color-view-controller.js]: done

File: emperor/support_files/js/color-view-controller.js (hunk 12 above)
Comment (squirrelo) [inline emperor/support_files/js/color-view-controller.js]: done

File: emperor/support_files/js/color-view-controller.js (hunk 15 above)
Comment (squirrelo) [inline emperor/support_files/js/color-view-controller.js]: The two passes is to do the validation that we do indeed have enough values to do the scaling. Not sure it can be done without two passes.

File: emperor/support_files/js/color-view-controller.js (hunk 13 above)
Comment (squirrelo) [inline emperor/support_files/js/color-view-controller.js]: Done.

File: emperor/support_files/js/color-view-controller.js (hunk 14 above)
Comment (squirrelo) [inline emperor/support_files/js/color-view-controller.js]: done

File: emperor/support_files/js/color-view-controller.js (hunk 17 above)
Comment (squirrelo) [inline emperor/support_files/js/color-view-controller.js]: done

File: emperor/support_files/js/color-view-controller.js (hunk 16 above)
Comment (squirrelo) [inline emperor/support_files/js/color-view-controller.js]: vestigial from refactor. Removed.

File: tests/javascript_tests/test_color_view_controller.js (hunk 19 above)
Comment (squirrelo) [inline tests/javascript_tests/test_color_view_controller.js]: No actual functionality was changed save that error throwing, so the tests made sure it still works the same. I can add helper function tests though.

File: emperor/support_files/js/color-view-controller.js (hunk 3 above)
Comment (ElDeveloper) [inline emperor/support_files/js/color-view-controller.js]: I see, but I don't think we should do that in this case.

File: emperor/support_files/js/color-view-controller.js (hunk 4 above)
Comment (ElDeveloper) [inline emperor/support_files/js/color-view-controller.js]: If I recall correctly, from their documentation, the grid assumes the
size of the containing div, so we just have to make this div the size of
the available space, which is what should already be happening AFAIK.

File: emperor/support_files/js/color-view-controller.js (hunk 15 above)
Comment (ElDeveloper) [inline emperor/support_files/js/color-view-controller.js]: One way to do it would be to do a single scan where we search and count
things, if we could also sort as we are doing this, that would be
exactly what we want. Maybe we can sketch something out.

File: tests/javascript_tests/test_color_view_controller.js (hunk 19 above)
Comment (ElDeveloper) [inline tests/javascript_tests/test_color_view_controller.js]: I see. I think that suggestion is good, could probably mean just
re-purposing some of the already existing tests, and making sure all
paths are covered.

**Comment by squirrelo** [comment]
Comments addressed. Fixed the shrinking SlickGrid issue and turned the colorbar 90 degrees to allow longer values:

screen shot 2016-05-09 at 13 32 36

**Comment by ElDeveloper** [comment]
Thanks for making these changes @squirrelo, I noticed a few things though. The sidebar is getting an odd size when you switch over to continuous values, there should not be a scrollbar, but there is, no idea why:

screen shot 2016-05-09 at 11 10 32 pm

Also I noticed the grid and bar are overlapping in some cases, the solution for this would be to use an HTML spacer or something like that:

screen shot 2016-05-09 at 11 11 17 pm

Finally, one minor thing that I only noticed when I looked at a real dataset 👀, the low small values of the color bar are at the top and the high values at the bottom, should have noticed before, but what we want is likely low values on the bottom, high values on top.

**Comment by ElDeveloper** [comment]
@squirrelo, I just noticed something else. I tried saving the setting for a plot where I had checked the "Continuous Values" checkbox, and set a different color for the missing values, and when I tried to load the settings, I got two errors:

screen shot 2016-05-09 at 11 21 21 pm

screen shot 2016-05-09 at 11 21 23 pm

Let me know if you need help reproducing this problem!

**Comment by squirrelo** [comment]
Thanks for catching those. Fixed now, and ready to go.

File: emperor/support_files/js/color-view-controller.js [hunk 20]
```diff
@@ -277,42 +394,50 @@ define([
...
+      // Get the current SlickGrid data and update with the saved color
+      var data = this.bodyGrid.getData();
+      data[0].value = json.data['Non-numeric values'];
+      this.setPlottableAttributes(decompViewDict, json.data['Non-numeric values'], data[0].plottables);
+
+    }
+    else {
+      var data = decompViewDict.setCategory(json.data, this.setPlottableAttributes, json.category);
+    }
+    console.log(data);
```
Comment (ElDeveloper) [inline emperor/support_files/js/color-view-controller.js]: Can you remove this line?

File: emperor/support_files/js/color-view-controller.js [hunk 21]
```diff
@@ -277,42 +394,50 @@ define([
...
+    this.setSlickGridDataset(data);
   }
 
-  /**
-   *
-   * Retrieve a discrete color.
+    /**
```
Comment (ElDeveloper) [inline emperor/support_files/js/color-view-controller.js]: Can you move this line back to how it was before?

**Comment by ElDeveloper** [comment]
@squirrelo, the grid gets resized oddly. To reproduce, scale continuous values, then switch tabs, switch back turn off the scaling and then turning it on:

screen shot 2016-05-10 at 3 05 18 pm

Note that we can't just hide the scrollbars, we need to actually fix this so the size of the table is correct. Also, why is the colorbar using so much vertical space?

**Comment by squirrelo** [comment]
Should be fixed now.

**Comment by ElDeveloper** [comment]
@squirrelo, this is working great, works flawlessly, the only odd thing is the display of the grid and the colorbar:

screen shot 2016-05-12 at 9 54 08 pm

Note how the grid is clipped, this is more evident when changing colors, not sure what's going on, some light debugging on Chrome seems to hint at the fact that the grid is inferring a different size (not 100% sure), no idea what might be going on.

Also why is there a space between the colorbar and the non-numeric grid? With those two things in, I think this is 👍 💯

**Comment by squirrelo** [comment]
I can't recreate the grid clipping, so unsure how to fix. The SVG is not obeying the height percentage given when in the notebook, but work as a standard index file. Will try to figure that out.

**Comment by ElDeveloper** [comment]
When I saw this problem, I was running emperor in the Jupyter notebook
and switching between tabs.

Thanks!

Yoshiki.

**Comment by squirrelo** [comment]
Issues should be addressed now.

File: emperor/support_files/js/color-view-controller.js [hunk 22]
```diff
@@ -187,77 +219,162 @@ define([
    * @param {scaled} Whether to use a scaled colormap or equidistant colors for
    * each value. Default is false.
    *
+   * @return {colors} The object containing the hex colors keyed to each sample
```
Comment (ElDeveloper) [inline emperor/support_files/js/color-view-controller.js]: I don't think we are allowed to have two `@return` tags, are we?

Comment (squirrelo) [inline emperor/support_files/js/color-view-controller.js]: We are, it's `@returns` thats bad.

File: emperor/support_files/js/color-view-controller.js [hunk 23]
```diff
@@ -187,77 +219,162 @@ define([
...
-        stopColors.push(interpolator(s).hex());
+  /**
+   *
+   * Retrieve a scaled color set with scale bar.
+   *
+   * @param {values} list of objects to generate a color for, usually a
+   * category in a given metadata column.
+   * @param {map} string, name of the discrete color map to use.
+   *
+   * @return {colors} The object containing the hex colors keyed to each sample
```
Comment (ElDeveloper) [inline emperor/support_files/js/color-view-controller.js]: Same comment regarding the `@return` tag.

File: emperor/support_files/js/color-view-controller.js [hunk 24]
```diff
@@ -187,77 +219,162 @@ define([
...
+   *
+   * Retrieve an interpolatd color set.
+   *
+   * @param {values} list of objects to generate a color for, usually a
+   * category in a given metadata column.
+   * @param {map} string, name of the discrete color map to use.
+   *
+   * @return {colors} The object containing the hex colors keyed to each sample
+   *
+   * Defaults to use viridis colormap if there's no map passed in.
```
Comment (ElDeveloper) [inline emperor/support_files/js/color-view-controller.js]: I think this should be part of the `@param {map}`.

File: emperor/support_files/js/color-view-controller.js [hunk 25]
```diff
@@ -187,77 +219,162 @@ define([
...
+   *
+   */
+  ColorViewController.getInterpolatedColors = function(values, map) {
+    map = map || 'viridis';
+    map = chroma.brewer[map];
+
+    var total = values.length;
+    interpolator = chroma.bezier([map[0], map[3], map[4], map[5], map[8]]);
```
Comment (ElDeveloper) [inline emperor/support_files/js/color-view-controller.js]: this is global, I think.

File: emperor/support_files/js/view-controller.js [hunk 26]
```diff
@@ -84,9 +84,9 @@ define([
     this.$header = $('<div name="emperor-view-controller-header"></div>');
     this.$header.css('margin', '0 auto');
     this.$header.css('width', '100%');
-
```
Comment (ElDeveloper) [inline emperor/support_files/js/view-controller.js]: Any reason why this line is removed?

File: emperor/support_files/js/view-controller.js [hunk 27]
```diff
@@ -84,9 +84,9 @@ define([
...
     this.$body = $('<div name="emperor-view-controller-body"></div>');
     this.$body.css('margin', '0 auto');
+
```
Comment (ElDeveloper) [inline emperor/support_files/js/view-controller.js]: Any reason why this line is added?

File: emperor/support_files/js/color-view-controller.js [hunk 28]
```diff
@@ -174,90 +181,209 @@ define([
...
-   * This function will generate a list of coloring values depending on the
-   * coloring scheme that the system is currently using (discrete or
-   * continuous).
    */
   ColorViewController.getColorList = function(values, map, discrete, scaled) {
-    var colors = {}, numColors = values.length - 1, counter = 0;
-    var interpolator, min, max;
-    var scaled = scaled || false;
+    var colors = {}, gradientSVG;
+    scaled = scaled || false;
```
Comment (mortonjt) [inline emperor/support_files/js/color-view-controller.js]: I'm just curious - do you really need this line?

**Comment by ElDeveloper** [comment]
🚢 :shipit:

File: emperor/support_files/js/color-view-controller.js [hunk 29]
```diff
@@ -174,90 +181,209 @@ define([
...
 
     if (_.findWhere(ColorViewController.Colormaps, {id: map}) === undefined){
       throw new Error("Could not find " + map + " as a colormap.");
     }
 
-    // 1 color and continuous coloring should return the first element of the
-    // map
-    if (numColors === 0 && discrete === false) {
+    // 1 color and continuous coloring should return the first element in map
+    if (values.length == 1 && discrete === false) {
```
Comment (mortonjt) [inline emperor/support_files/js/color-view-controller.js]: `value.length == 1` or `values.length === 1`?

File: emperor/support_files/js/color-view-controller.js (hunk 28 above)
Comment (squirrelo) [inline emperor/support_files/js/color-view-controller.js]: yes, it makes the coloring default to not showing scaled bar if called, which should happen.

File: emperor/support_files/js/color-view-controller.js (hunk 29 above)
Comment (squirrelo) [inline emperor/support_files/js/color-view-controller.js]: either works in this case since we are calling length.

File: emperor/support_files/js/color-view-controller.js [hunk 30]
```diff
@@ -174,90 +181,209 @@ define([
...
+   * @param {String} [map] name of the color map to use, see
+   * ColorViewController.Colormaps Defaults to use qiime discrete colors if
+   * there's no map passed in.
+   *
+   * @return {Object} [colors] The object containing the hex colors keyed to
+   * each sample
+   *
+   */
+  ColorViewController.getDiscreteColors = function(values, map) {
+    map = map || 'discrete-coloring-qiime';
```
Comment (mortonjt) [inline emperor/support_files/js/color-view-controller.js]: 👍 pretty slick

File: emperor/support_files/js/color-view-controller.js [hunk 31]
```diff
@@ -174,90 +181,209 @@ define([
...
+
+    if (map == 'discrete-coloring-qiime') {
```
Comment (mortonjt) [inline emperor/support_files/js/color-view-controller.js]: `map == 'discrete-coloring-qiime'` or `map === 'discrete-coloring-qiime'`?

File: emperor/support_files/js/color-view-controller.js (hunk 29 above)
Comment (mortonjt) [inline emperor/support_files/js/color-view-controller.js]: sounds good.

File: emperor/support_files/js/color-view-controller.js (hunk 28 above)
Comment (mortonjt) [inline emperor/support_files/js/color-view-controller.js]: gotcha.
